ocrd_utils/ocrd_utils/os.py

```diff
@@ -90,15 +94,36 @@ def list_all_resources(executable):
     processor_path_var = '%s_PATH' % executable.replace('-', '_').upper()
     if processor_path_var in environ:
         for processor_path in environ[processor_path_var].split(':'):
-            if isdir(processor_path):
-                candidates += list(scandir(processor_path))
-    datadir = join(XDG_DATA_HOME, 'ocrd-resources', executable)
-    if isdir(datadir):
-        candidates += list(scandir(datadir))
-    systemdir = join('/usr/local/share/ocrd-resources', executable)
-    if isdir(systemdir):
-        candidates += list(scandir(systemdir))
-    return [x.path for x in candidates]
+            if Path(processor_path).is_dir():
+                candidates += Path(processor_path).iterdir()
+    datadir = Path(XDG_DATA_HOME, 'ocrd-resources', executable)
+    if datadir.is_dir():
+        candidates += datadir.iterdir()
+    systemdir = Path('/usr/local/share/ocrd-resources', executable)
+    if systemdir.is_dir():
+        candidates += systemdir.iterdir()
+    # recurse once
+    for parent in candidates:
+        if parent.is_dir():
+            candidates += parent.iterdir()
+    if is_dir:
+        candidates = [x for x in candidates if x.is_dir()]
+    else:
+        candidates = [x for x in candidates if not x.is_dir()]
```

Comment:
Like I said, I would find it better to check the file type on a candidate by candidate basis (so we don't need to assume file and directory parameters never co-occur in a tool's params – as they do in ocrd-calamari-recognize's `checkpoint` vs `checkpoint_dir` params). The candidates could be returned on the condition that `x.exists()` here, and in `show_resource`, one would only have to check `is_dir` again to discern the two cases.

Reply:
Couldn't we look into the tool json once, and get back with a boolean tuple `dir_res, file_res` (so either or both or none can be true), and then list accordingly? (Since you have already separated the question of whether to recurse...)

Reply:
You mean "only output directories iff all file parameters are directories"? This would effectively mean the status quo (list all files and folders) if a procesor has both file and directory resources. Which affects only ocrd_calamari AFAICT.

But sure, it's at least less hacky than checking if any parameter is a directory parameter and potentially not listing relevant resources.

Reply:
No, I meant
- "output directories under resource paths if any file parameters are directories"
- "output files under resource paths if any file parameters are files"
(either of which could be empty or none or both)

Reply:
Gotcha 229925d5 replaces `are_processor_resources_directories` with `get_processor_resource_types` which returns a tuple `has_dirs, has_files` and then we filter based on that after getting the list of candidates.

This works nicely but has the drawback that due to the selection criteria (not all file parameters have `content-type`), it breaks e.g. `ocrd-tesserocr-recognize -L`. Adding a `content-type` to `ocrd-tesserocr-recognize`'s `model` parameter fixes that.

But since we will have to do work on the various ocrd-tool.json to decentralize resource lists anyway, I think we should really fix the processors in this regard, not work around it in core.

Reply:
Agreed!